src/main/kotlin/org/rust/ide/inspections/RsNeedlessLifetimesInspection.kt

```diff
@@ -0,0 +1,240 @@
+                else -> false  // already elided, different named lifetimes or something static going on
+            }
+        }
+        else -> false
+    }
+}
+
+private fun allowedLifetimesFrom(lifetimeParameters: List<RsLifetimeParameter>): Set<ReferenceLifetime> {
+    val allowedLifetimes = HashSet<ReferenceLifetime>()
+    lifetimeParameters
+        .filter { it.bounds.isEmpty() }
+        .mapNotNull { it.name }
+        .map { ReferenceLifetime.Named(it) }
+        .toCollection(allowedLifetimes)
+    allowedLifetimes.add(ReferenceLifetime.Unnamed)
+    allowedLifetimes.add(ReferenceLifetime.Static)
+    return allowedLifetimes
+}
+
+private fun lifetimesFromBounds(bounds: List<RsLifetime>): List<ReferenceLifetime> =
+    bounds.map { it.name }.filterIsInstance<ReferenceLifetime.Named>()
+
+private fun countUniqueLifetimes(lifetimes: List<ReferenceLifetime>): Int = HashSet(lifetimes).size
+
+/**
+ * Are any lifetimes mentioned in the `where` clause? If yes, we don't try to reason about elision.
+ */
+private fun hasWhereLifetimes(whereClause: RsWhereClause?): Boolean {
+    if (whereClause == null) return false
+    for (predicate in whereClause.wherePredList) {
+        if (predicate.lifetime != null) return true
+        // a predicate like F: Trait or F: for<'a> Trait<'a>
+
+        // check the type F, it may not contain LT refs
+        val ty = predicate.typeReference?.type
+        if (ty?.hasReEarlyBounds == true) return true
+
+        // if the bounds define new lifetimes, they are fine to occur
+        val boundLifetimeParams = predicate.forLifetimes?.lifetimeParameterList.orEmpty()
+        val allowedLifetimes = allowedLifetimesFrom(boundLifetimeParams)
+        // now walk the bounds
+        val collector = LifetimesCollector()
+        predicate.typeParamBounds?.polyboundList?.map { it.bound }?.forEach { it.accept(collector) }
+        if (collector.abort) return false
+        // and check that all lifetimes are allowed
+        if (collector.lifetimes.any { !allowedLifetimes.contains(it) }) return false
+    }
+    return false
+}
+
+/**
+ * The lifetime of a &-reference.
+ */
+private sealed class ReferenceLifetime {
+    object Unnamed : ReferenceLifetime()
+    object Static : ReferenceLifetime()
+    data class Named(val name: String) : ReferenceLifetime()
+}
+
+private class LifetimesCollector : RsVisitor() {
+    val lifetimes = mutableListOf<ReferenceLifetime>()
+    var abort: Boolean = false
+        private set
+
+    override fun visitLifetime(o: RsLifetime) = record(o)
+
+//    override fun visitTypeReference(o: RsTypeReference) {
+//        if (o.refLikeType?.lifetime.name.isElided) {
+//            record(null)
+//        }
+//    }
+
+//    override fun visitPath(o: RsPath) {
+//        collectAnonymousLifetimes(o)
+//        // TODO: handle existential types
+//    }
+
+    override fun visitElement(o: RsElement) {
+        if (!abort && o !is RsItemElement /* ignore nested items */) {
```

Comment:
I think it's better to invert this condition like this: 
```kotlin
if (abort || o is RsItemElement) return
o.childrenOfType<RsElement>().forEach { it.accept(this) }
```

It's always easier to read conditions without negations